lang/query_test.go

```diff
@@ -0,0 +1,175 @@
+package lang_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	_ "github.com/influxdata/flux/builtin"
+
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/lang"
+	"github.com/influxdata/flux/memory"
+	"github.com/pkg/errors"
+)
+
+func runQuery(script string) (flux.Query, error) {
+	program, err := lang.Compile(script, time.Unix(0, 0))
+	if err != nil {
+		return nil, err
+	}
+	q, err := program.Start(context.Background(), &memory.Allocator{})
+	if err != nil {
+		return nil, err
+	}
+	return q, nil
+}
+
+var validScript = `
+import "csv"
+
+data = "
+#datatype,string,long,long,string
+#group,false,false,false,true
+#default,_result,,,
+,result,table,value,tag
+,,0,10,a
+,,0,10,a
+,,1,20,b
+,,1,20,b
+,,2,30,c
+,,2,30,c
+,,3,40,d
+,,3,40,d
+"
+
+csv.from(csv: data) |> yield(name: "res")`
+
+func TestQuery_Results(t *testing.T) {
+	q, err := runQuery(validScript)
+	if err != nil {
+		t.Fatal(errors.Wrap(err, "unexpected error while creating query"))
+	}
+
+	// gather counts
+	var resCount, tableCount, totRows int
+	for res := range q.Results() {
+		resCount++
+		if err := res.Tables().Do(func(tbl flux.Table) error {
+			tableCount++
+			return tbl.Do(func(cr flux.ColReader) error {
+				tags := cr.Strings(1)
+				for i := 0; i < tags.Len(); i++ {
+					totRows++
+				}
+				return nil
+			})
+		}); err != nil {
+			t.Fatal(errors.Wrap(err, "unexpected error while iterating over tables"))
+		}
+	}
+
+	// com,pare expected counts
+	if resCount != 1 {
+		t.Errorf("got %d results instead of %d", resCount, 1)
+	}
+	if tableCount != 4 {
+		t.Errorf("got %d tables instead of %d", tableCount, 4)
+	}
+	if totRows != 8 {
+		t.Errorf("got %d rows instead of %d", totRows, 8)
+	}
+
+	// release query resources
+	q.Done()
+
+	if q.Err() != nil {
+		t.Fatal(errors.Wrap(q.Err(), "unexpected error from query execution"))
+	}
+}
+
+func TestQuery_Stats(t *testing.T) {
+	t.Skip("stats are updated by the controller, running a standalone query won't update them")
```

Comment:
Hmm, it occurs to me that because `lang.Query` needs to use an `execute.Executor`, this is hard to test.  One solution might be to add a method that lets clients decide which executor to use when running the program, in this interface:
https://github.com/influxdata/flux/blob/master/lang/compiler.go#L208
```go
type DependenciesAwareProgram interface {
	SetExecutorDependencies(execute.Dependencies)
	SetLogger(logger *zap.Logger)
        SetExecutor(e execute.Executor)
}
```
Then you could pass in a `mock.Executor` that updates the stats.

Reply:
I am a little reluctant to add code that is only used for tests, but it seems like the benefits outweigh the cost here.

Reply:
I see your point. But, if I do it for this test, isn't it the same if the query itself returns the stats it wants to return?

I say this, because, for "real" queries, it is the controller that sets statistics.  
Moreover, when it comes to the executor, the query has already been compiled and planned, so some stats can't be updated (or they should be update with random values).

Reply:
Hmm.  This is becoming a design discussion, that should involve our teammates.

Both the `Query` interface and the `ResultIterator` that wraps it are tricky, especially with regard to canceling and statistics, and so I wanted to ensure that we have coverage against breaking things, as well as a good example for clients to follow.

I think you're right that since the controller is the thing that updates the stats, maybe it's really the controller we want to test (or refactor like you mention).

Reply:
I guess what I'm saying is if you want to merge this as-is, and then we can file other issues where we are missing test coverage, that would be fine.

Reply:
Yeah, I think this point is a bit tricky and worth a new issue and relative PR.